config/exec_resource.go

```diff
@@ -0,0 +1,184 @@
+			return "", fmt.Errorf("unsupported interpreter %q", execR.GetInterpreter())
+		}
+		name := filepath.Join(tmpDir, "")
+		_, err := util.WriteFile(strings.NewReader(execR.GetScript()), "", name, 0644)
+		if err != nil {
+			return "", err
+		}
+
+	case *agentendpointpb.ApplyConfigTask_OSPolicy_Resource_ExecResource_Exec_File:
+		if execR.GetFile().GetLocalPath() != "" {
+			return execR.GetFile().GetLocalPath(), nil
+		}
+		switch {
+		case execR.GetFile().GetGcs().GetObject() != "":
+			name = path.Base(execR.GetFile().GetGcs().GetObject())
+		case execR.GetFile().GetRemote().GetUri() != "":
+			name = path.Base(execR.GetFile().GetRemote().GetUri())
+		default:
+			return "", fmt.Errorf("unsupported File %v", execR.GetFile())
+		}
+		name = filepath.Join(tmpDir, name)
+		if _, err := downloadFile(ctx, name, execR.GetFile()); err != nil {
+			return "", err
+		}
+	default:
+		return "", fmt.Errorf("unrecognized Source type for FileResource: %q", execR.GetSource())
+	}
+
+	return name, nil
+}
+
+func (e *execResource) validate(ctx context.Context) (*ManagedResources, error) {
+	tmpDir, err := ioutil.TempDir("", "osconfig_exec_resource_")
+	if err != nil {
+		return nil, fmt.Errorf("failed to create temp dir: %s", err)
+	}
+	e.tempDir = tmpDir
+
+	e.validatePath, err = e.download(ctx, e.GetValidate())
+	if err != nil {
+		return nil, err
+	}
+
+	e.enforcePath, err = e.download(ctx, e.GetEnforce())
+	if err != nil {
+		return nil, err
+	}
+
+	return nil, nil
+}
+
+func (e *execResource) run(ctx context.Context, name string, execR *agentendpointpb.ApplyConfigTask_OSPolicy_Resource_ExecResource_Exec) ([]byte, []byte, int, error) {
+	var cmd string
+	var args []string
+	switch execR.GetInterpreter() {
+	case agentendpointpb.ApplyConfigTask_OSPolicy_Resource_ExecResource_Exec_NONE:
+		cmd = name
+	case agentendpointpb.ApplyConfigTask_OSPolicy_Resource_ExecResource_Exec_SHELL:
+		if runtime.GOOS == "windows" {
+			cmd = name
+		} else {
+			args = append([]string{name})
+			cmd = "/bin/sh"
+		}
+	case agentendpointpb.ApplyConfigTask_OSPolicy_Resource_ExecResource_Exec_POWERSHELL:
+		if runtime.GOOS != "windows" {
+			return nil, nil, 0, fmt.Errorf("interpreter %q can only be used on Windows systems", execR.GetInterpreter())
+		}
+		args = append([]string{"-File", name})
+		cmd = "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\PowerShell.exe"
+	default:
+		return nil, nil, 0, fmt.Errorf("unsupported interpreter %q", execR.GetInterpreter())
+	}
+
+	stdout, stderr, err := runner.Run(ctx, exec.Command(cmd, args...))
+	code := -1
+	if err != nil {
+		if v, ok := err.(*exec.ExitError); ok {
+			code = v.ExitCode()
+		}
+	}
+	return stdout, stderr, code, err
+}
+
+func (e *execResource) checkState(ctx context.Context) (inDesiredState bool, err error) {
+	// For validate we expect an exit code of 100 for "correct state" and 101 for "incorrect state".
+	stdout, stderr, code, err := e.run(ctx, e.validatePath, e.GetValidate())
+	switch code {
+	case -1:
```

Comment:
Is this a code that the customer returns or is this what the OS returns?

Reply:
The way this logic works is if e.run returns code -1 that means it was an error running exec (like file not found or something) so we just log the error. this is a workaround for checking err != nil as we can't do that directly because all non 0 return codes return an exec error.